crates/binjs_io/src/entropy/baseline.rs

```diff
@@ -0,0 +1,229 @@
+//! baseline dictionary. It is possible to use only a baseline dictionary, but experiments
+//! shows a ~50% file size increase (see https://github.com/binast/binjs-ref/issues/249).
+//! It is also possible to use only a sampling dictionary, at the cost of extensibility,
+//! but experiments do not indicate any advantage to doing so, except perhaps faster
+//! initialization.
+
+use entropy::Dictionary;
+use io::statistics::Instances;
+
+use binjs_meta::spec::{self, Spec};
+use binjs_shared::{FieldName, InterfaceName, SharedString};
+
+use std::borrow::Borrow;
+use std::collections::{HashMap, HashSet};
+use std::rc::Rc;
+
+use itertools::Itertools;
+
+type IOPath = binjs_shared::ast::Path<
+    InterfaceName,
+    (
+        /* child index */ usize,
+        /* field name */ FieldName,
+    ),
+>;
+
+/// Build a baseline dictionary from a spec.
+///
+/// `depth` is the depth of the dictionary, i.e. the length of paths.
+pub fn build(depth: usize, spec: &Spec) -> Dictionary<Instances> {
+    let mut builder = BaselineDictionaryBuilder::new(depth, spec);
+    builder.start();
+    builder.done()
+}
+
+/// A structure dedicated to building a baseline dictionary.
+struct BaselineDictionaryBuilder<'a> {
+    /// The dictionary currently being constructed.
+    dictionary: Dictionary<Instances>,
+
+    /// The specs from which to extract the dictionary.
+    spec: &'a Spec,
+
+    /// The name of the special `null` interface, used to represent the `null` case of `attribute Foo?`.
+    null_name: InterfaceName,
+}
+impl<'a> BaselineDictionaryBuilder<'a> {
+    /// Create a new baseline dictionary builder for a given depth and spec.
+    pub fn new(depth: usize, spec: &'a Spec) -> Self {
+        let null_name = InterfaceName::from_rc_string(spec.get_null_name().to_rc_string().clone());
+        debug!(target: "baseline", "Starting baseline with depth {depth}, null = \"{null}\", root = \"{root}\"",
+            null = null_name.as_str(),
+            root = spec.get_root_name().to_str(),
+            depth = depth);
+
+        debug!(target: "baseline", "Roots: [{:?}]\n",
+            spec.resolved_sums_of_interfaces_by_name()
+                .get(spec.get_root_name())
+                .unwrap()
+                .iter()
+                .map(|node_name| InterfaceName::from_rc_string(node_name.to_rc_string().clone()))
+                .format(", "));
+
+        BaselineDictionaryBuilder {
+            spec,
+            null_name,
+            dictionary: Dictionary::new(depth, 0),
+        }
+    }
+
+    fn seed_one_field(
+        &mut self,
+        path: &IOPath,
+        type_spec: &spec::TypeSpec,
+        or_null: bool,
+        one_level_paths_by_interface_name: &mut HashMap<InterfaceName, HashSet<IOPath>>,
+    ) {
+        use self::spec::NamedType;
+        use self::spec::TypeSpec;
+        match type_spec {
+            TypeSpec::Array { ref contents, .. } => {
+                // Entering an array doesn't affect the path.
```

Comment:
can you explain a bit more what this comment implicitly means?
this sounds like some operation is omitted here, but not clear what it is.